### ISSUE TITLE
Angular Universal (SSR) not working when this plugin is used

### ISSUE THREAD

**Issue by harm-less** [jaydenseric/apollo-upload-client#153]
Hey,

I've been struggling to figure out an extremely weird bug that seems to be caused somehow by this plugin.

I'm using Angular, GraphQL/Apollo and this upload plugin all on their latest version. I was trying to set up Angular Universal but somehow certain route resolvers didn't resolve properly and I didn't get the full pre-rendered page back. After painstaking debugging, I noticed it only happened when a GraphQL call was made (non of them had anything to do with uploading though). So step by step I was disabling things that were not essential to load a certain page in my project. It turned out that the once I replaced the `createUploadLink` method with a normal, basic `httpLink.create` it all started to work.

My current working code is using this:
```ts
const STATE_KEY = makeStateKey('apollo.state');

const isBrowser = this.transferState.hasKey<any>(STATE_KEY);
const options: HttpOptions = { uri: `${environment.api}/graphql` };
const http = isBrowser ? createUploadLink(options) : httpLink.create(options);

this.apollo.create({
  link: http,
  cache: new InMemoryCache()
});
```
As you can see I simply do not allow the `createUploadLink` to be used when `isBrowser` is not true. This is not a fix but a workaround of course, but it works for me because I don't need to upload anything while prerendering.

So in other words, I'm pretty sure this plugin does something to mess up the resolvers, I just can't figure out what makes that happen.

To go a bit more into detail about my findings:
1. It doesn't matter if my resolver returned an `Observable` or a `Promise`, both stopped before the call completed
2. The Apollo calls complete anyway as I received the data. It just "finalized" the SSR request before it was going to use the data it received from the server.

It's hard for me to pinpoint where this issue comes from exactly, but it might be very useful information for people using this plugin and Angular Universal. I hope @jaydenseric will have some kind of epiphany based on this description 😉

**Comment by jaydenseric** [comment]
I'm a little confused by this issue.

You're not using [`apollo-angular`](https://github.com/apollographql/apollo-angular) and [`apollo-angular-link-http`](https://github.com/apollographql/apollo-angular/tree/master/packages/apollo-angular-link-http), which has built-in support for file uploads?

> somehow certain route resolvers didn't resolve properly and I didn't get the full pre-rendered page back

What does `didn't resolve properly` mean, what exactly is the error?

> I'm pretty sure this plugin does something to mess up the resolvers

This is a client library, there are no resolvers involved on the client.

> It doesn't matter if my resolver returned an `Observable` or a `Promise`, both stopped before the call completed

I have not heard of returning an `Observable` for GraphQL server resolvers. And what does "stopped" mean? An error?

**Comment by harm-less** [comment]
I do remember reading something about it being supported by the official Apollo package, so I guess I should give that a go instead.

> What does "didn't resolve properly" mean, what exactly is the error?

Yeah, it's really hard to explain without a demo perhaps

> I have not heard of returning an Observable for GraphQL server resolvers. And what does "stopped" mean? An error?

I meant an observable in the form of something like this: `Observable<ApolloQueryResult<MyQuery>>`

But I see now that explaining the problem I encountered without a video or demo is going to be really hard.
I think it's best I look into the build-in support for upload in `apollo-angular` and close this issue. The conversation just might help someone in the future.
